packages/api/src/controllers/animalTypeController.js

```diff
@@ -34,6 +37,43 @@ const animalTypeController = {
       }
     };
   },
+
+  addFarmAnimalType() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+      try {
+        const { farm_id } = req.headers;
+        const { type } = req.body;
+
+        const record = await baseController.existsInTable(trx, AnimalTypeModel, {
+          type,
+          farm_id,
+          deleted: false,
+        });
+
+        if (record) {
+          await trx.rollback();
+          return res.status(409).send();
+        } else {
+          const result = await baseController.postWithResponse(
+            AnimalTypeModel,
+            { type, farm_id },
+            req,
+            {
+              trx,
+            },
+          );
```

Comment:
To simplify this logic, maybe we could start giving knex's `onConflict` a try? https://knexjs.org/guide/query-builder.html#onconflict

Essentially we can tell it to insert except if certain attributes match in which case we want it to ignore and do nothing.

Reply:
I think makes sense for when our goal is upsert (like the code you flagged in the notification badge PR https://github.com/LiteFarmOrg/LiteFarm/pull/3054) but don't we want the conflict response returned (rather than a silent fail) in this case?

I'm not sure because we also haven't really consumed this error status on the frontend -- usually we've aimed to give a validation error before we even hit the API in the case of existing records -- so this would be more about what you see when using the API directly.

Reply:
That's a good point! I do like that the API response is more explicit in this scenario and indicates the conflict, so I'm on board with keeping it this way. The rest looks great to me, I haven't tested it out yet but I'll do that in a bit!